modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,358 @@
+* `Nothing` type accepts no value, but it can be assigned to all the types. This type is not frequently used explicitly, but it is used by the type inference algorithm.
+
+=== Union Type
+
+The `Union` type is used to compose types. The syntax to define a `Union` type is:
+
+----
+TypeA | TypeB | ...
+----
+
+The following example defines a variable with a constraint that accepts a value of type `String` or `Number`:
+
+----
+var age: String | Number = if (payload.allStrings) "32" else 32
+----
+
+A common pattern is to use `Null` type with a `Union` type to specify that a type accepts `null` as a value. In the following example, the variable `a` allows the expression `payload.foo` to return `null`.
+
+----
+var a: String| Null = payload.foo
+----
+
+
+=== Intersection Type
+
+The `Intersection` type intersects `Object` types. In this case, the intersection works as the concatenation (`++`) of object types.
+
+_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._
+
+The syntax for `Intersection` type is:
+----
+TypeA & TypeB & ...
+----
+
+In the following example, the `Intersection` concatenates the two `Object` types, resulting in the type `{name: String, lastName: String}`. The type is an open object that can have more key-value pairs. The variable accepts the value assigned to it:
+
+----
+var a: {name: String} & {lastName: String} = {name: "John", lastName: "Smith", age: 34}
+----
+
+In the case of closed objects, it returns the concatenation of the object types but resuls in a closed object. +
+In the following example, the intersection results in the type `{|name: String, lastName: String|}`, which throws
+an exception because a closed object does not accept an `Object` value if there are additional fields in the object (field `age`):
+
+----
+var a: {|name: String|} & {|lastName: String|} = {name: "John", lastName: "Smith", age: 34}
+----
+
+[[literal-types]]
+=== Literal Types
+
+A literal type represents exactly one value. For example, the String value `"foo"` can be represented with the type `"foo"`.
+
+_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._
+
+The following literal types have been added to the type system:
+
+* `String` literal types
+* `Number` literal types
+* `Boolean` literal types
+
+You can use literal types with `Union` types to declare a type as a finite set of allowed values. For example, the following type declarations are aliases of `Union` and literal types:
+
+----
+type Weekdays = "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday"
+type Days = Weekdays | "Saturday" | "Sunday"
+----
+
+In the following example, the variable only accepts a value of the literal types `404` and `500`. The type system ensures that the variable is one of those literal values:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+var errorStatusCode: 404 | 500 = payload.statusCode
+----
+
+Function overloading enables you to define different behaviors based on the input argument's value:
```

Comment:
Is this still about Literal types?
If not, we need a transition or new heading.